### PR TITLE
Rewrite General RNA-Seq pipeline to Python

### PR DESCRIPTION
**Pull request by OtonicarJan**
## Checklist

* [x] Update CHANGELOG.rst for each commit separately:
  * Pay attention to write entries under the "Unreleased" section.
  * Mark all breaking changes as "**BACKWARD INCOMPATIBLE:**" and put them
    before non-breaking changes.
  * If a commit modifies a feature listed under "Unreleased" section,
    it might be sufficient to modify the existing CHANGELOG entry from previous
    commit(s).
* [x] Bump the process version:
  * **MAJOR version (first number)**: Backward incompatible changes (changes
    that break the api/interface). Examples: renaming the input/output, adding
    mandatory input, removing input/output...
  * **MINOR version (middle number)**: add functionality or changes in a
    backwards-compatible manner. Examples: add output field, add non-mandatory
    input parameter, use a different tool that produces same results...
  * **PATCH version (last number)**: changes/bug fixes that do not affect
    the api/interface. Examples: typo fix, change/add warning messages...
* [x] All inputs are used in process.
* [x] All output fields have a value assigned to them.

### CODE REVIEW

File: resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py [hunk 1]
```diff
@@ -0,0 +1,802 @@
...
+    process_type = "data:workflow:rnaseq:featurecounts:qc"
+    category = "Pipeline"
+
+    class Input:
+        """Input fields."""
+
+        class Preprocessing:
+            """Preprocessing with BBDuk."""
+
+            reads = DataField(
```
Comment (jkokosar) [inline resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py:56]: I suggest to exclude the required inputs (reads, genome, etc) from the corresponding parent groups and leave them on the top level. List the groups of non-required inputs after the required inputs.  Boolean field `show_advanced` can then be omitted.

File: resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py [hunk 2]
```diff
@@ -0,0 +1,802 @@
...
+                    "do not soft-clip. Extend5pOfRead1: fully extend only the 5p of the read1, all "
+                    "other ends: local alignment. Extend5pOfReads12: fully extend only the 5' of the "
+                    "both read1 and read2, all other ends use local alignment.",
+                    default="Local",
+                )
+
+            class TwoPassOptions:
+                """Two-pass mapping options."""
+
+                two_pass_mode = BooleanField(
```
Comment (jkokosar) [inline resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py:320]: This is defined but not used anywhere. STAR two-pass is most probably not relevant for this pipeline.

File: resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py [hunk 3]
```diff
@@ -0,0 +1,802 @@
...
+            annotation = DataField(
+                "annotation",
+                label="Annotation",
+                description="GTF and GFF3 annotation formats are supported",
+            )
+            show_advanced = BooleanField(
+                label="Advanced options",
+                default=False,
+            )
+            assay_type = StringField(
```
Comment (jkokosar) [inline resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py:400]: I'm in favour of exposing `assay_type` and related `cdna_index` inputs on the top-level, as this is a critical parameter that must not be selected in the wrong way (it must match the library type of the dataset being analysed). If the setting is hidden in the advanced section of the inputs, it is easy to get this wrong.

File: resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py [hunk 4]
```diff
@@ -0,0 +1,802 @@
...
+            )
+
+        class QualityControl:
+            """Quality control."""
+
+            rrna_reference = DataField(
+                "index:star",
+                label="Indexed rRNA reference sequence",
+                description="Reference sequence index prepared by STAR aligner indexing tool.",
+                required=False,
```
Comment (jkokosar) [inline resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py:522]: This is a required input.

File: resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py [hunk 5]
```diff
@@ -0,0 +1,802 @@
...
+            )
+            globin_reference = DataField(
+                "index:star",
+                label="Indexed Globin reference sequence",
+                description="Reference sequence index prepared by STAR aligner indexing tool.",
+                required=False,
```
Comment (jkokosar) [inline resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py:528]: This is a required input.

File: resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py (hunk 4 above)
Comment (jkokosar) [reply resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py:522]: It should also be on the top level of the inputs.

File: resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py [hunk 6]
```diff
@@ -0,0 +1,802 @@
...
+                else -1,
+                "quality_trim": "r",
+                "trim_quality": inputs.preprocessing.trim_quality,
+                "quality_encoding_offset": inputs.preprocessing.quality_encoding_offset,
+                "ignore_bad_quality": inputs.preprocessing.ignore_bad_quality,
+                "maxns": inputs.preprocessing.maxns,
+            },
+        }
+
+        if inputs.preprocessing.reads.type == "data:reads:fastq:single:":
```
Comment (jkokosar) [inline resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py:587]: Use `startswith()` for this check, as the pipeline should fit all subtypes of the reads objects, e.g. "data:reads:fastq:single:bbduk:" or "data:reads:fastq:paired:bbduk:"

File: resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py [hunk 7]
```diff
@@ -0,0 +1,802 @@
...
+            preprocessing = Data.create(
+                process=BioProcess.get_latest(slug="bbduk-single"),
+                input=input_bbduk,
+            )
+        elif inputs.preprocessing.reads.type == "data:reads:fastq:paired:":
```
Comment (jkokosar) [inline resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py:592]: Add also the `else` statement, where the workflow is terminated if the expected input type is not provided.

File: resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py [hunk 8]
```diff
@@ -0,0 +1,802 @@
...
+                "quality_trim": "r",
+                "trim_quality": inputs.preprocessing.trim_quality,
+                "quality_encoding_offset": inputs.preprocessing.quality_encoding_offset,
+                "ignore_bad_quality": inputs.preprocessing.ignore_bad_quality,
+                "maxns": inputs.preprocessing.maxns,
+            },
+        }
+
+        if inputs.preprocessing.reads.type == "data:reads:fastq:single:":
+            preprocessing = Data.create(
```
Comment (jkokosar) [inline resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py:588]: if you assign just the process slug to a variable here (`process_slug = "bbduk-single"`), you can use just one `Data.create(slug = process_slug, input = input_bbduk)` statement outside the `if elif else` block.

File: resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py [hunk 9]
```diff
@@ -0,0 +1,802 @@
...
+            "aligned_reads": alignment,
+            "n_reads": inputs.quantification.n_reads,
+            "assay_type": inputs.quantification.assay_type,
+            "annotation": inputs.quantification.annotation,
+            "feature_class": inputs.quantification.feature_class,
+            "feature_type": inputs.quantification.feature_type,
+            "id_attribute": inputs.quantification.id_attribute,
+        }
+
+        if inputs.quantification.cdna_index:
```
Comment (jkokosar) [inline resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py:734]: Add an additional check here for the `cdna_index` input. If the `cdna_index` is not given when the `assay_type` is set to auto, fail the workflow. Add this check to the beginning of the workflow, so no individual processes are triggered if this state is detected.

File: resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py [hunk 10]
```diff
@@ -0,0 +1,802 @@
...
+            "reads": preprocessing,
+            "n_reads": inputs.downsampling.n_reads,
+            "advanced": {
+                "seed": inputs.downsampling.advanced.seed,
+                "fraction": inputs.downsampling.advanced.fraction,
+                "two_pass": inputs.downsampling.advanced.two_pass,
+            },
+        }
+
+        if inputs.preprocessing.reads.type == "data:reads:fastq:single:":
```
Comment (jkokosar) [inline resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py:753]: Use `startswith()`

File: resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py [hunk 11]
```diff
@@ -0,0 +1,708 @@
...
+            preprocessing,
+            alignment,
+            downsampling,
+            quantification,
+            alignment_qc_rrna,
+            alignment_qc_globin,
+            idxstats,
+        ]
+
+        inputs_multiqc = {
```
Comment (romunov) [inline resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.py:703]: Above chunk `multiqc_inputs` could be incorporated into this dict without an intermediate step as `multiqc_inputs` is not used anywhere else.

Renaming `inputs_multiqc` to `input_multiqc` would make it more in line with the rest of the input variable names used in this process.

**Review by romunov — APPROVED** [review]
Yuge work, nicely done!

File: docs/CHANGELOG.rst [hunk 12]
```diff
@@ -33,6 +33,9 @@ Changed
   ``workflow-bbduk-star-htseq``).
 - **BACKWARD INCOMPATIBLE:** Unify Feature autocomplete and search
   endpoints into a single endpoint
+- **BACKWARD INCOMPATIBLE:** Rewrite processes
```
Comment (jkokosar) [inline docs/CHANGELOG.rst:36]: Move to the unreleased section

**Pull request merged**